Sources/SwiftParser/Lexer/UnicodeScalarExtensions.swift

```diff
@@ -179,20 +173,22 @@ extension Unicode.Scalar {
       return Unicode.Scalar(curByte)
     }
 
-    // Read the number of high bits set, which indicates the number of bytes in
-    // the character.
-    let encodedBytes = (~(UInt32(curByte) << 24)).leadingZeroBitCount
-
-    // If this is 0b10XXXXXX, then it is a continuation character.
-    if encodedBytes == 1 || !Unicode.Scalar(curByte).isStartOfUTF8Character {
+    // If this is not the start of a UTF8 character,
+    // then it is either a continuation byte or an invalid UTF8 code point.
+    if !curByte.isStartOfUTF8Character {
       // Skip until we get the start of another character.  This is guaranteed to
       // at least stop at the nul at the end of the buffer.
-      while let peeked = peek(), !Unicode.Scalar(peeked).isStartOfUTF8Character {
+      while let peeked = peek(), !peeked.isStartOfUTF8Character {
         _ = advance()
       }
       return nil
     }
 
+    // Read the number of high bits set, which indicates the number of bytes in
+    // the character.
+    let encodedBytes = (~curByte).leadingZeroBitCount
+    precondition(encodedBytes >= 2 && encodedBytes <= 4)
```

Comment:
Could you add a doc comment why this precondition holds? Something like

```swift
// We have a multi-byte UTF-8 scalar.
// Single-byte UTF-8 scalars are handled at the start of the function by checking `curByte < 0x80`.
// `isStartOfUTF8Character` guaranteed that the `curByte` has 2 to 4 leading ones.
```

Just to clarify that we should never hit this precondition if running the lexer on invalid UTF-8.

Reply:
Added! 
Your words accurately articulate the rationale for the precondition, so no modifications are needed.